File: backend/src/Sites/Site.cs
using Stronzflix.Media;

namespace Stronzflix.Sites
{
    public abstract class Site
    {
        public string Name { get; private set; }
        public string Url { get; private set; }

        public Site(string name, string url)
        {
            this.Name = name;
            this.Url = url;
        }

        public abstract Result[] Search(string query);

        public abstract Title GetTitle(string url);

        public abstract string GetSource(string url);
    }
}

File: lib/backend/site.dart
import 'package:stronzflix/backend/media.dart';

abstract class Site {

    final String name;
    final String url;

    Site(this.name, this.url) {
        this.prepare().then((_) => Site._registry[name] = this);
    }

    Future<void> prepare() async {}
    Future<List<SearchResult>> search(String query);
    Future<Title> getTitle(SearchResult result);

    static final Map<String, Site> _registry = {};
    static Site? get(String name) => _registry[name];
}
